File: tel.ts
export const TELPattern = '^\\+?(\\d{1,3})?[-.\\s]?(\\(?\\d{1,4}\\)?)?[-.\\s]?\\d{1,4}[-.\\s]?\\d{1,4}[-.\\s]?\\d{1,9}$';
export const reTEL = new RegExp(TELPattern);

function stripTEL(tel: string): string {
    return tel.trim().replaceAll(/\D/gi,'');
}

export function normalizeTEL(tel: string): string {
    const bareTel = stripTEL(tel);
    switch (bareTel.length) {
        case 7:
            return `${bareTel.substring(0,3)}-${bareTel.substring(3,)}`
        case 10:
            return `(${bareTel.substring(0,3)}) ${bareTel.substring(3,6)}-${bareTel.substring(6,)}`;
        case 11:
            return `+00${bareTel.substring(0,1)} (${bareTel.substring(1,4)}) ${bareTel.substring(4,7)}-${bareTel.substring(7,)}`;
        case 12:
            return `+0${bareTel.substring(0,2)} (${bareTel.substring(2,5)}) ${bareTel.substring(5,8)}-${bareTel.substring(8,)}`;
        case 13:
            return `+${bareTel.substring(0,3)} (${bareTel.substring(3,6)}) ${bareTel.substring(6,9)}-${bareTel.substring(9,)}`;
        default:
            return tel;
    }
}

export function validateTEL(tel: string): boolean {
    return reTEL.test(tel);
}

File: tel.go
package metadatatools

import (
	"regexp"
	"strings"
)

// TELPattern is a permissive regular expression pattern for telephone
// numbers in a variety of common formats.
const TELPattern = `^\+?(\d{1,3})?[-.\s]?(\(?\d{1,4}\)?)?[-.\s]?\d{1,4}[-.\s]?\d{1,4}[-.\s]?\d{1,9}$`

// reTEL is the compiled form of TELPattern.
var reTEL = regexp.MustCompile(TELPattern)

// stripTEL trims whitespace from tel and removes any remaining
// non-digit characters.
func stripTEL(tel string) string {
	return reNonDigit.ReplaceAllString(strings.TrimSpace(tel), "")
}

// NormalizeTEL formats tel based on the number of digits it contains.
// If the digit count does not match a recognized length, tel is
// returned unchanged.
func NormalizeTEL(tel string) string {
	bare := stripTEL(tel)
	switch len(bare) {
	case 7:
		return bare[0:3] + "-" + bare[3:]
	case 10:
		return "(" + bare[0:3] + ") " + bare[3:6] + "-" + bare[6:]
	case 11:
		return "+00" + bare[0:1] + " (" + bare[1:4] + ") " + bare[4:7] + "-" + bare[7:]
	case 12:
		return "+0" + bare[0:2] + " (" + bare[2:5] + ") " + bare[5:8] + "-" + bare[8:]
	case 13:
		return "+" + bare[0:3] + " (" + bare[3:6] + ") " + bare[6:9] + "-" + bare[9:]
	default:
		return tel
	}
}

// ValidateTEL validates tel against TELPattern.
func ValidateTEL(tel string) bool {
	return reTEL.MatchString(tel)
}
